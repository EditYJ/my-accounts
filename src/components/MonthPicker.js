/* eslint-disable jsx-a11y/anchor-is-valid */
import PropTypes from "prop-types";
import React from "react";

import { addZero, range, viewActive } from "../utility";

class MonthPicker extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isOpen: false,
      selectedYear: this.props.year
      // selectedMonth: this.props.month
    };
  }

  componentDidMount() {
    document.addEventListener("click", this.handleClick, false);
  }
  componentWillUnmount() {
    document.removeEventListener("click", this.handleClick, false);
  }

  handleClick = event => {
    if (this.node.contains(event.target)) {
      return;
    }
    this.setState({
      isOpen: false
    });
  };

  // 切换下拉显隐
  toggleDropdown = event => {
    event.preventDefault();
    this.setState({
      isOpen: !this.state.isOpen
    });
  };

  toggleYearLis = (event, yearNumber) => {
    event.preventDefault();
    this.setState({
      selectedYear: yearNumber
    });
  };

  toggleMonthLis = (event, monthNumber) => {
    event.preventDefault();
    this.setState({
      isOpen: false
      // selectedMonth: monthNumber
    });
    this.props.onChange(this.state.selectedYear, monthNumber);
  };

  render() {
    const { year, month } = this.props;
    const { isOpen } = this.state;
    const monthRange = range(12, 1);
    const yearRange = range(9, -4).map(number => number + year);

    return (
      <div
        className="dropdown month-picker-component"
        ref={ref => (this.node = ref)}
      >
        <h6>选择月份</h6>
        <button
          className="btn btn-secondary dropdown-toggle"
          onClick={this.toggleDropdown}
        >
          {`${year}年 ${addZero(month)}月`}
        </button>
        {isOpen && (
          <div className="dropdown-menu" style={{ display: "block" }}>
            {/* <div
              className="row"
              style={{ display: "flex", "justify-content": "center" }}
            >
              <a
                className="query-all"
                href="#"
                onClick={event => this.toggleMonthLis(event, true)}
              >
                查询全部
              </a>
            </div> */}
            <div className="row">
              <div className="col border-right years-range">
                {yearRange.map((yearNumber, index) => {
                  return (
                    <a
                      key={index}
                      href="#"
                      className={viewActive(
                        yearNumber,
                        this.state.selectedYear
                      )}
                      onClick={event => this.toggleYearLis(event, yearNumber)}
                    >
                      {yearNumber}年
                    </a>
                  );
                })}
              </div>
              <div className="col months-range">
                {monthRange.map((monthNumber, index) => {
                  return (
                    <a
                      key={index}
                      href="#"
                      className={viewActive(monthNumber, month)}
                      onClick={event => this.toggleMonthLis(event, monthNumber)}
                    >
                      {addZero(monthNumber)}月
                    </a>
                  );
                })}
              </div>
            </div>
          </div>
        )}
      </div>
    );
  }
}

MonthPicker.protoType = {
  year: PropTypes.number.isRequired,
  month: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired
};

MonthPicker.defaultProps = {
  onChange: () => {}
};

export default MonthPicker;
