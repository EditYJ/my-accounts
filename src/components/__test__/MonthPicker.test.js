import React from "react";
import ReactDOM from "react-dom";
import { mount } from "enzyme";
import MonthPicker from "../MonthPicker";

const props = {
  year: 2018,
  month: 8,
  onChange: jest.fn()
};

let wrapper;

describe("test MonthPicker component", () => {
  beforeEach(() => {
    wrapper = mount(<MonthPicker {...props} />);
  });

  it("should render the component to match snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("测试默认状态-按钮上渲染年和月是否正确，下拉框是否显示", () => {
    const text = wrapper
      .find(".dropdown-toggle")
      .first()
      .text();
    expect(text).toEqual("2018年 08月");
    expect(wrapper.find(".dropdown-menu").length).toEqual(0);
    expect(wrapper.state("isOpen")).toEqual(false);
    expect(wrapper.state("selectedYear")).toEqual(props.year);
  });

  it("测试点击以后状态-下拉框是否显示,年和月的列表是否正确等", () => {
    wrapper.find(".dropdown-toggle").simulate("click");
    expect(wrapper.state("isOpen")).toEqual(true);
    expect(wrapper.find(".dropdown-menu").length).toEqual(1);
    expect(wrapper.find(".years-range .dropdown-item").length).toEqual(9);
    expect(wrapper.find(".months-range .dropdown-item").length).toEqual(12);
    expect(wrapper.find(".years-range .dropdown-item.active").text()).toEqual(
      "2018年"
    );
    expect(wrapper.find(".months-range .dropdown-item.active").text()).toEqual(
      "08月"
    );

    expect(
      wrapper
        .find(".years-range .dropdown-item")
        .first()
        .text()
    ).toEqual(`${props.year - 4}年`);

    expect(
      wrapper
        .find(".months-range .dropdown-item")
        .first()
        .text()
    ).toEqual(`01月`);
  });

  it("点击年和月份以后是否触发正确的回调", () => {
    wrapper.find(".dropdown-toggle").simulate("click");
    wrapper
      .find(".years-range .dropdown-item")
      .first()
      .simulate("click");
    expect(
      wrapper
        .find(".years-range .dropdown-item")
        .first()
        .hasClass("active")
    ).toEqual(true);
    expect(wrapper.state("selectedYear")).toEqual(2014);
    wrapper
      .find(".months-range .dropdown-item")
      .first()
      .simulate("click");
    expect(wrapper.state("isOpen")).toEqual(false);
    expect(props.onChange).toHaveBeenCalledWith(2014, 1);
  });

  it("点击document应该关闭展开的日历选择器", () => {
    let eventMap = {};
    document.addEventListener = jest.fn((event, cb) => {
      eventMap[event] = cb
    });
    const wrapper = mount(<MonthPicker {...props} />);
    wrapper.find(".dropdown-toggle").simulate("click");
    expect(wrapper.state("isOpen")).toEqual(true);
    expect(wrapper.find(".dropdown-menu").length).toEqual(1);

    eventMap.click({
      target: ReactDOM.findDOMNode(wrapper.instance())
    })
    expect(wrapper.state("isOpen")).toEqual(true);
    eventMap.click({
      target: document
    })
    expect(wrapper.state("isOpen")).toEqual(false);
  });
});
