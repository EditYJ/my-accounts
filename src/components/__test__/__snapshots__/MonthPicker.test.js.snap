// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MonthPicker component should render the component to match snapshot 1`] = `
<MonthPicker
  month={8}
  onChange={[MockFunction]}
  year={2018}
>
  <div
    className="dropdown month-picker-component"
  >
    <h6>
      选择月份
    </h6>
    <button
      className="btn btn-secondary dropdown-toggle"
      onClick={[Function]}
    >
      2018年 08月
    </button>
  </div>
</MonthPicker>
`;
